go-controller/pkg/ovnwebhook/podadmission.go

```diff
@@ -0,0 +1,138 @@
+package ovnwebhook
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kubevirt"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	"golang.org/x/exp/maps"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	listers "k8s.io/client-go/listers/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type checkPodAnnot func(nodeLister listers.NodeLister, v annotationChange, pod *corev1.Pod, nodeName string) error
+
+// interconnectPodAnnotationChecks holds annotations allowed for ovnkube-node:<nodeName> users in IC environments
+var interconnectPodAnnotations = map[string]checkPodAnnot{
+	util.OvnPodAnnotationName: func(nodeLister listers.NodeLister, v annotationChange, pod *corev1.Pod, nodeName string) error {
+		// Ignore kubevirt pods with live migration, the IP can cross node-subnet boundaries
+		if kubevirt.IsPodLiveMigratable(pod) {
+			return nil
+		}
+
+		podAnnot, err := util.UnmarshalPodAnnotation(map[string]string{util.OvnPodAnnotationName: v.value}, types.DefaultNetworkName)
+		if err != nil {
+			return err
+		}
+		node, err := nodeLister.Get(nodeName)
+		if err != nil {
+			return fmt.Errorf("could not get info on node %s from client: %w", nodeName, err)
+		}
+
+		subnets, err := util.ParseNodeHostSubnetAnnotation(node, types.DefaultNetworkName)
+		if err != nil {
+			return err
+		}
+		for _, ip := range podAnnot.IPs {
+			found := false
+			for _, subnet := range subnets {
+				if subnet.Contains(ip.IP) {
+					found = true
+				}
+			}
+			if !found {
+				return fmt.Errorf("%s does not belong to %s node", ip, nodeName)
```

Comment:
what about ensuring someone does not change the IP? We can also check if the IP matches the pod status IPs. If pod status IPs are not nil, they should match the IPs in this annotation or we should error.

Reply:
Although this sounds like a good idea I am not sure if this is something we want to do now. This adds some extra computations and complexity but a compromised node could still potentially break pods running on it. The main objective of this change is to make sure that a specific ovnkube-node is not able to affect workloads running on different nodes.

Reply:
OK we can defer to later.